cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DatastoreImpl.java

```diff
@@ -1525,95 +1525,83 @@ boolean isOpen() {
-        for(Map.Entry<String, String> e : revisions.entries()) {
-            current.put(e.getKey(), e.getValue());
-            // the query uses below (see revsDiffBatch())
-            // `multimap.size() + multimap.keySet().size()` placeholders
-            // and SQLite has limit on the number of placeholders on a single query.
-            if(current.size() + current.keySet().size() >= size) {
-                partitions.add(current);
-                current = HashMultimap.create();
-            }
-        }
-
-        if(current.size() > 0) {
-            partitions.add(current);
-        }
-
-        return partitions;
-    }
-
     /**
-     * Removes revisions present in the datastore from the input map.
+     * Checks the supplied collection of revisions for the given document ID and returns a
+     * collection containing only those entries that are missing from the database.
      *
-     * @param revisions an multimap from document id to set of revisions. The
-     *                  map is modified in place for performance consideration.
+     * @param db    the database to get the revisions from
+     * @param docId the doc ID to check
+     * @param revs  the rev IDs to check
+     * @return collection of rev IDs not present in the database
      */
-    void revsDiffBatch(SQLDatabase db,Multimap<String, String> revisions) throws DatastoreException {
+    Collection<String> revsDiffBatch(SQLDatabase db, String docId, Collection<String> revs)
+            throws DatastoreException {
+
+        // Consider all missing to start
+        Set<String> missingRevs = new HashSet<String>(revs);
 
         final String sql = String.format(
-                "SELECT docs.docid, revs.revid FROM docs, revs " +
-                "WHERE docs.doc_id = revs.doc_id AND docs.docid IN (%s) AND revs.revid IN (%s) " +
-                "ORDER BY docs.docid",
-                DatabaseUtils.makePlaceholders(revisions.keySet().size()),
-                DatabaseUtils.makePlaceholders(revisions.size()));
-
-        String[] args = new String[revisions.keySet().size() + revisions.size()];
-        String[] keys = revisions.keySet().toArray(new String[revisions.keySet().size()]);
-        String[] values = revisions.values().toArray(new String[revisions.size()]);
-        System.arraycopy(keys, 0, args, 0, revisions.keySet().size());
-        System.arraycopy(values, 0, args, revisions.keySet().size(), revisions.size());
+                "SELECT revs.revid FROM docs, revs " +
+                        "WHERE docs.doc_id = revs.doc_id AND docs.docid = %s AND revs.revid IN " +
+                        "(%s) ",
+                DatabaseUtils.makePlaceholders(1),
+                DatabaseUtils.makePlaceholders(revs.size()));
+
+        String[] args = new String[1 + revs.size()];
+        args[0] = docId;
+        // Copy the revisions into the end of the args array
+        System.arraycopy(revs.toArray(new String[revs.size()]), 0, args, 1, revs.size());
```

Comment:
We still need to pass the docId in these args to fill in the `?` placeholder. We could use `String[] args = revs.toArray()` if we substitute the docId directly into the selection string instead of `?` but in that case I think we need to do some string escaping as well.